hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenUpdateNftsHandler.java

```diff
@@ -81,8 +89,25 @@ public void preHandle(@NonNull final PreHandleContext context) throws PreCheckEx
         final var tokenStore = context.createStore(ReadableTokenStore.class);
         final var token = tokenStore.get(op.tokenOrElse(TokenID.DEFAULT));
         if (token == null) throw new PreCheckException(INVALID_TOKEN_ID);
-        if (token.hasMetadataKey()) {
-            context.requireKey(token.metadataKeyOrThrow());
+
+        final var nftStore = context.createStore(ReadableNftStore.class);
+        if (serialNumbersInTreasury(
+                token.treasuryAccountIdOrThrow(), op.serialNumbers(), nftStore, token.tokenIdOrThrow())) {
+            if (token.hasMetadataKey() && token.hasSupplyKey()) {
+                context.requireKey(oneOf(token.metadataKeyOrThrow(), token.supplyKeyOrThrow()));
+            } else if (token.hasMetadataKey()) {
+                context.requireKey(token.metadataKeyOrThrow());
+            } else if (token.hasSupplyKey()) {
+                context.requireKey(token.supplyKeyOrThrow());
+            } else {
+                throw new PreCheckException(TOKEN_IS_IMMUTABLE);
```

Comment:
We can add a response code something like `TOKEN_HAS_NO_METADTA_OR_SUPPLY_KEY`. We usually throw `TOKEN_IS_IMMUTABLE` when there is no `adminKey` on token

Reply:
Added new response code to protobuf. Opening a PR for the protobuf branch.